### PR TITLE
BCDA 3578 Added v2 endpoint scaffolding 

### PR DESCRIPTION
**Pull request by PatKelsh**
### Fixes [BCDA-3578](https://jira.cms.gov/browse/BCDA-3578)

Creating the API scaffolding for the v2 data request endpoints. The new endpoints will use the same logic as their v1 counterparts, just using the v2 endpoint. Logic for the v2 endpoints will be done in a later ticket. Created a new feature flag to limit access to the new endpoints.

### Proposed Changes

* Created a new v2 api.go file 
* Updated router to define the new endpoints
* New feature flag to hide the new endpoints by default

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### CODE REVIEW

File: bcda/web/api_v2.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	log "github.com/sirupsen/logrus"
+
+	"github.com/CMSgov/bcda-app/bcda/database"
+	"github.com/CMSgov/bcda-app/bcda/models"
+	"github.com/CMSgov/bcda-app/bcda/responseutils"
+	"github.com/CMSgov/bcda-app/bcda/servicemux"
+	"github.com/CMSgov/bcda-app/bcda/utils"
+)
+
+// var (
```
Comment (mtrang1263) [inline bcda/web/api_v2.go:26]: What do you think about moving these all to a separate directory/package? So we would have `bcda/web/v2/api.go`?

That way we keep all of the resources/constants decoupled and we support _semantic import versioning_ which is referenced in the official [Go docs](https://blog.golang.org/versioning-proposal).

We'd want to make the handlers exported but then we can drop the v2 naming suffix.

File: bcda/web/router.go [hunk 2]
```diff
@@ -34,6 +34,14 @@ func NewAPIRouter() http.Handler {
 		r.With(auth.RequireTokenAuth, auth.RequireTokenJobMatch).Get(m.WrapHandler("/jobs/{jobID}", jobStatus))
 		r.Get(m.WrapHandler("/metadata", metadata))
 	})
+
+	if utils.GetEnvBool("VERSION_2_ENDPOINT_ACTIVE", false) {
+		r.Route("/api/v2", func(r chi.Router) {
+			r.With(auth.RequireTokenAuth, ValidateBulkRequestHeaders).Get(m.WrapHandler("/Patient/$export", bulkPatientRequestV2))
```
Comment (mtrang1263) [inline bcda/web/router.go:40]: If we decide on moving out the v2 into a separate package, we can use named imports to help with the package name collision.

So we could have the following:
```
import (
...
webV2 github.com/CMSgov/bcda-app/bcda/web/v2
...
)
```

And then we would use `webV2.BulkPatientRequestV2`

File: bcda/web/router_test.go [hunk 3]
```diff
@@ -178,7 +177,32 @@ func (s *RouterTestSuite) TestCoverageExportRoute() {
 
 	res = s.getAPIRoute("/api/v1/Groups/new/$export?_type=Coverage")
 	assert.Equal(s.T(), http.StatusNotFound, res.StatusCode)
+}
+
+func (s *RouterTestSuite) TestV2EndpointsDisabled() {
```
Comment (msnook) [inline bcda/web/router_test.go:182]: Should we test the job status endpoint too?

File: bcda/web/router_test.go [hunk 4]
```diff
@@ -178,7 +177,32 @@ func (s *RouterTestSuite) TestCoverageExportRoute() {
...
+	os.Setenv("VERSION_2_ENDPOINT_ACTIVE", "false")
+	s.apiRouter = NewAPIRouter()
 
+	res := s.getAPIRoute("/api/v2/Patient/$export")
+	assert.Equal(s.T(), http.StatusNotFound, res.StatusCode)
+	res = s.getAPIRoute("/api/v2/Group/all/$export")
+	assert.Equal(s.T(), http.StatusNotFound, res.StatusCode)
+}
+
+func (s *RouterTestSuite) TestV2EndpointsEnabled() {
```
Comment (msnook) [inline bcda/web/router_test.go:195]: Should we test the job status endpoint too?

File: bcda/web/router.go [hunk 5]
```diff
@@ -34,6 +34,14 @@ func NewAPIRouter() http.Handler {
...
+			r.With(auth.RequireTokenAuth, ValidateBulkRequestHeaders).Get(m.WrapHandler("/Group/{groupId}/$export", bulkGroupRequestV2))
```
Comment (msnook) [inline bcda/web/router.go:41]: Should we include the job status endpoint?

File: bcda/web/v2/api.go [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+		if err = qc.Enqueue(j); err != nil {
+			log.Error(err)
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Processing, "")
+			responseutils.WriteError(oo, w, http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func check429(jobs []models.Job, types []string, w http.ResponseWriter) ([]string, bool) {
```
Comment (msnook) [inline bcda/web/v2/api.go:279]: Wouldn't this be the same across versions?  How do we implement `DRY`?

File: bcda/web/router.go [hunk 7]
```diff
@@ -34,6 +35,14 @@ func NewAPIRouter() http.Handler {
 		r.With(auth.RequireTokenAuth, auth.RequireTokenJobMatch).Get(m.WrapHandler("/jobs/{jobID}", jobStatus))
 		r.Get(m.WrapHandler("/metadata", metadata))
 	})
+
+	if utils.GetEnvBool("VERSION_2_ENDPOINT_ACTIVE", false) {
+		r.Route("/api/v2", func(r chi.Router) {
+			r.With(auth.RequireTokenAuth, ValidateBulkRequestHeaders).Get(m.WrapHandler("/Patient/$export", v2.BulkPatientRequest))
+			r.With(auth.RequireTokenAuth, ValidateBulkRequestHeaders).Get(m.WrapHandler("/Group/{groupId}/$export", v2.BulkGroupRequest))
```
Comment (msnook) [inline bcda/web/router.go:42]: Job status endpoint?

Comment (PatKelsh) [reply bcda/web/router.go:42]: The ticket says only to move over the patient and group endpoints

File: bcda/web/api_test.go [hunk 8]
```diff
@@ -1178,16 +1178,6 @@ func (s *APITestSuite) TestJobStatusFailed() {
 
 	assert.Equal(s.T(), http.StatusInternalServerError, s.rr.Code)
 
-	var oo fhirmodels.OperationOutcome
```
Comment (msnook) [inline bcda/web/api_test.go:1181]: Hmm, do you need to pull `master` into this branch?  Isn't this from a previous PR?

File: bcda/web/v2/api.go [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+		}
+	}
+	if len(unworkedTypes) == 0 {
+		return nil, false
+	} else {
+		return unworkedTypes, true
+	}
+}
+
+func validateRequest(r *http.Request) ([]string, *fhirmodels.OperationOutcome) {
```
Comment (msnook) [inline bcda/web/v2/api.go:312]: How do we consolidate these duplicated functions between `v2` and `v1`?   Can they be broken out into a common area, or does it make sense to have a `v1` version and a `v2` version of these common functions?

File: bcda/web/v2/api_test.go [hunk 10]
```diff
@@ -0,0 +1,1534 @@
+package v2
```
Comment (msnook) [inline bcda/web/v2/api_test.go:1]: Looks like a lot of `DRY` here.  I wonder if we can have a single `api_test.go` which covers both `v1` tests and `v2` tests with common code?  Or are they different enough that separate explicit tests are warranted?  I worry about maintaining duplicate code here.

File: bcda/api/requests.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+		}
+	}
+	if len(unworkedTypes) == 0 {
+		return nil, false
+	} else {
+		return unworkedTypes, true
+	}
+}
+
+func ValidateRequest(r *http.Request) ([]string, *fhirmodels.OperationOutcome) {
```
Comment (msnook) [inline bcda/api/requests.go:218]: I'm wondering about function visibility here... should this be an exported function?

File: bcda/api/requests.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+		if err = qc.Enqueue(j); err != nil {
+			log.Error(err)
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Processing, "")
+			responseutils.WriteError(oo, w, http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func Check429(jobs []models.Job, types []string, w http.ResponseWriter) ([]string, bool) {
```
Comment (msnook) [inline bcda/api/requests.go:181]: I'm wondering about function visibility here... should this be an exported function?

File: bcda/api/v2/api.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/CMSgov/bcda-app/bcda/models"
+	"github.com/CMSgov/bcda-app/bcda/responseutils"
+	"github.com/CMSgov/bcda-app/bcda/utils"
+)
+
+const (
+	groupAll = "all"
+)
+
+func init() {
```
Comment (mtrang1263) [inline bcda/api/v2/api.go:22]: Can we move this `init()` method into bcda/api/requests.go? We want to have the initialization step shared between the v1/v2 API impls.

File: bcda/api/v2/api.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+Data export job has completed successfully. The response body will contain a JSON object providing metadata about the transaction.
+swagger:response completedJobResponse
+*/
+// nolint
+type CompletedJobResponse struct {
+	// in: body
+	Body bulkResponseBody
+}
+
+type bulkResponseBody struct {
```
Comment (mtrang1263) [inline bcda/api/v2/api.go:128]: Can we pull these two structs to request.go?

**Review by mtrang1263 — APPROVED** [review]
LGTM!

**Review by msnook — APPROVED** [review]
👍

**Pull request merged**